planner/core/logical_plan_builder.go

```diff
@@ -2962,62 +2982,139 @@ func (b *PlanBuilder) buildWindowFunctionFrame(spec *ast.WindowSpec, orderByItem
+			schema.Append(&expression.Column{
+				ColName:      model.NewCIStr(fmt.Sprintf("%d_window_%d", window.id, schema.Len())),
+				UniqueID:     b.ctx.GetSessionVars().AllocPlanColumnID(),
+				IsReferenced: true,
+				RetType:      desc.RetTp,
+			})
+		}
+		window.WindowFuncDescs = descs
+		window.SetChildren(np)
+		window.SetSchema(schema)
+		p = window
 	}
+	return p, windowMap, nil
+}
 
-	needFrame := aggregation.NeedFrame(expr.F)
+func extractWindowFuncs(fields []*ast.SelectField) []*ast.WindowFuncExpr {
+	extractor := &WindowFuncExtractor{}
+	for _, f := range fields {
+		n, _ := f.Expr.Accept(extractor)
+		f.Expr = n.(ast.ExprNode)
+	}
+	return extractor.windowFuncs
+}
+
+func (b *PlanBuilder) handleDefaultFrame(spec *ast.WindowSpec, name string) (*ast.WindowSpec, bool) {
+	needFrame := aggregation.NeedFrame(name)
 	// According to MySQL, In the absence of a frame clause, the default frame depends on whether an ORDER BY clause is present:
 	//   (1) With order by, the default frame is equivalent to "RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW";
 	//   (2) Without order by, the default frame is equivalent to "RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING",
 	//       which is the same as an empty frame.
-	if needFrame && expr.Spec.Frame == nil && len(orderBy) > 0 {
-		expr.Spec.Frame = &ast.FrameClause{
+	if needFrame && spec.Frame == nil && spec.OrderBy != nil {
+		newSpec := *spec
+		newSpec.Frame = &ast.FrameClause{
 			Type: ast.Ranges,
 			Extent: ast.FrameExtent{
 				Start: ast.FrameBound{Type: ast.Preceding, UnBounded: true},
 				End:   ast.FrameBound{Type: ast.CurrentRow},
 			},
 		}
+		return &newSpec, true
 	}
 	// For functions that operate on the entire partition, the frame clause will be ignored.
-	if !needFrame && expr.Spec.Frame != nil {
-		b.ctx.GetSessionVars().StmtCtx.AppendNote(ErrWindowFunctionIgnoresFrame.GenWithStackByArgs(expr.F, expr.Spec.Name.O))
-		expr.Spec.Frame = nil
-	}
-	frame, err := b.buildWindowFunctionFrame(&expr.Spec, orderBy)
-	if err != nil {
-		return nil, err
+	if !needFrame && spec.Frame != nil {
+		specName := spec.Name.O
+		b.ctx.GetSessionVars().StmtCtx.AppendNote(ErrWindowFunctionIgnoresFrame.GenWithStackByArgs(name, specName))
+		newSpec := *spec
+		newSpec.Frame = nil
+		return &newSpec, true
+	}
+	return spec, false
+}
+
+// groupWindowFuncs groups the window functions according to the window specification name.
+// TODO: We can group the window function by the definition of window specification.
+func (b *PlanBuilder) groupWindowFuncs(windowFuncs []*ast.WindowFuncExpr) (map[*ast.WindowSpec][]*ast.WindowFuncExpr, error) {
+	// updatedSpec is used to handle the specifications that have frame clause changed.
+	updatedSpec := make(map[string]*ast.WindowSpec)
+	groupedWindow := make(map[*ast.WindowSpec][]*ast.WindowFuncExpr)
+	for _, windowFunc := range windowFuncs {
+		if windowFunc.Spec.Name.L == "" {
+			windowFunc.Spec.Name = model.NewCIStr("<unnamed window>")
```

Comment:
This will make preapre && execute fail. We'd better use another way.

Reply:
Seems that this one is just moving code? If so, we can keep it in this pr and let https://github.com/pingcap/tidb/pull/9795 to fix it.

Reply:
Yes, just wait for which one merged first.